url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XUtils.java

```diff
@@ -0,0 +1,562 @@
+			return value;
+		XValue type = value.getType(runtime);
+		XObject obj = runtime.getObject(type);
+		XTypeData typeData = (XTypeData) obj.getData();
+		List<XValue> classes = typeData.getCRO();
+		for (XValue t : classes) {
+			if (t.getRaw(runtime, attr) != null) {
+				return t;
+			}
+		}
+		return null;
+	}
+
+	public static boolean isOverwritten(XRuntime runtime, XValue value,
+			String attr) {
+		XValue declarer = getDeclaring(runtime, value, attr, XValue.REF_NONE);
+		return declarer != null && declarer != runtime.getBaseType(OBJECT);
+	}
+
+	@SuppressWarnings("unchecked")
+	public static List<XValue> asList(XRuntime runtime, XValue pointer) {
+		XObject obj = runtime.getObject(pointer);
+		XObjectData data = obj == null ? null : obj.getData();
+		return data instanceof List ? (List<XValue>) data : null;
+	}
+
+	public static void putAtMapItem(XRuntime rt, XValue v1, String s, XValue v2) {
+		int i = s.indexOf('.');
+		int l = 0;
+		while (i != -1) {
+			String sub = s.substring(l, i);
+			XValue v3 = lookupTry(rt, v1, sub, XValue.REF_NONE);
+			if (v3 == null || v3 == XValueNull.NULL) {
+				v3 = rt.createMap(new HashMap<String, XValue>());
+				set(rt, v1, sub, v3, XValue.REF_NONE);
+			}
+			v1 = v3;
+			l = i + 1;
+			i = s.indexOf('.', l);
+		}
+		set(rt, v1, s.substring(l), v2, XValue.REF_NONE);
+	}
+
+	public static String getString(XRuntime rt, XValue value) {
+		XObjectData data = getData(rt, value);
+		return data instanceof XObjectDataString ? ((XObjectDataString) data)
+				.getString() : null;
+	}
+
+	@SuppressWarnings("unchecked")
+	public static <T> T getDataAs(XRuntime rt, XValue value, Class<T> c) {
+		XObjectData data = getData(rt, value);
+		return data != null && c.isAssignableFrom(data.getClass()) ? (T) data
+				: null;
+	}
+
+	public static XObjectData getData(XRuntime rt, XValue value) {
+		XObject obj = rt.getObject(value);
+		return obj == null ? null : obj.getData();
+	}
+
+	public static boolean isInstanceOf(XRuntime rt, XValue instance, XValue type) {
+		XValue instanceType = instance.getType(rt);
+		return isDerivedOf(rt, type, instanceType);
+	}
+
+	public static boolean isDerivedOf(XRuntime rt, XValue base, XValue type) {
+		XObject typeObj = rt.getObject(type);
+		XObject baseObj = rt.getObject(base);
+		if (typeObj == null || !typeObj.isType(rt) || baseObj == null
+				|| !baseObj.isType(rt))
+			return false;
+		return ((XTypeData) typeObj.getData()).getCRO().contains(base);
+	}
+
+	public static void rethrow(Throwable e) {
+		if (e instanceof RuntimeException) {
+			throw (RuntimeException) e;
+		} else if (e instanceof Error) {
+			throw (Error) e;
+		}
+		throw new RuntimeException(e);
+	}
+
+	public static XValue[] wrap(XRuntime rt, Object[] obj) {
+		XValue[] values = new XValue[obj.length];
+		for (int i = 0; i < obj.length; i++) {
+			values[i] = wrap(rt, obj[i]);
+		}
+		return values;
+	}
+
+	public static XValue wrap(XRuntime rt, Object obj) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.